assets/src/edit-story/components/dropTargets/withDropTarget.js

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useDrop } from 'react-dnd';
+
+/**
+ * Internal dependencies
+ */
+import { useEffect, useCallback, useState } from 'react';
+import StoryPropTypes from '../../types';
+import { MASKS } from '../../masks/index';
+import { useStory } from '../../app/story';
+
+const dropTargetStyles = {
+  transition: 'opacity 0.5s',
+  pointerEvents: 'visibleStroke',
+};
+
+const dropTargetContainerStyle = {
+  position: 'absolute',
+  top: 0,
+  left: 0,
+  width: '100%',
+  height: '100%',
+  pointerEvents: 'none',
+};
+
+function WithDropTarget({ element, children }) {
+  const {
+    actions: { updateElementById },
+  } = useStory();
+  const setProperties = useCallback(
+    (properties) => updateElementById({ elementId: element.id, properties }),
+    [element.id, updateElementById]
+  );
+
+  const [initialBackgroundImage, setInitialBackgroundImage] = useState(
+    element.backgroundImage
+  );
+
+  const [{ isOver, item }, drop] = useDrop({
+    accept: 'media',
+    drop: () => {
+      setInitialBackgroundImage(item.id);
+      setProperties({
+        backgroundImage: item.id,
+      });
+    },
+    collect: (monitor) => ({
+      isOver: monitor.isOver({ shallow: true }),
+      item: monitor.getItem(),
+    }),
+  });
+
+  // Previews
+  useEffect(() => {
+    if (!isOver || !item) {
+      // Reset background image when the element is out of the drop target
+      if (element.backgroundImage !== initialBackgroundImage) {
```

Comment:
Let's discuss this. An alternative model could be simply to swap the `Frame` and `Display` elements (or maybe even just `Display` is enough). We already have a transformer handler. The nice part about the current Wrapper->Display architecture is that we could simply supply a replacement element  and swap it on the fly. This could be important, b/c the correct default display is tricky and may require some special calculations that we do not want to repeat. Thus, structurally, the dropping will be a function - a function that takes the existing `element` and mutates it to the new replaced `element`. So it'd look like:

1. Execute drop function: `replacement = dropFunc(element)`
2. Push transform: `pushTransform({replacement})`
3. Change the display via state setter inside the `displayElement.js` via the `useTransform`.